src/main/java/com/vlkan/pubsub/PubsubSubscriber.java

```diff
@@ -0,0 +1,222 @@
+                            return delayEnabled
+                                    ? Mono.just(requestIndex).delayElement(config.getPullPeriod(), scheduler)
+                                    : Mono.just(requestIndex);
+                        })
+                        .flatMap(
+                                requestIndex -> pullAndConsumeAndAck(emptyBatchReceivedRef, requestIndex),
+                                config.getPullConcurrency()))
+                .subscribeOn(scheduler);
+    }
+
+    private Mono<PubsubSubscriberConsumption> pullAndConsumeAndAck(
+            AtomicBoolean emptyBatchReceivedRef,
+            BigDecimal requestIndex) {
+        return client
+                .pull(config.getProjectName(), config.getSubscriptionName(), pullRequest)
+                .filter(pullResponse -> {
+                    boolean emptyBatchReceived = pullResponse.getReceivedAckableMessages().isEmpty();
+                    emptyBatchReceivedRef.set(emptyBatchReceived);
+                    return !emptyBatchReceived;
+                })
+                .flatMap(this::consumeAndAck)
+                .transform(this::rateLimit)
+                // Warning! onErrorResume() should be placed after transform()
+                // calls so the latter can observe errors. [vyazici]
+                .onErrorResume(error -> {
+                    String errorMessage = String.format(
+                            "failed consuming message (projectName=%s, subscriptionName=%s, requestIndex=%s)",
+                            config.getProjectName(), config.getSubscriptionName(), requestIndex);
+                    LOGGER.error(errorMessage, error);
+                    return Mono.empty();
+                });
+    }
+
+    private Mono<PubsubSubscriberConsumption> consumeAndAck(PubsubPullResponse pullResponse) {
+        return Mono
+                .defer(() -> {
+
+                    // Discard empty batches.
+                    List<PubsubReceivedAckableMessage> receivedAckableMessages = pullResponse.getReceivedAckableMessages();
+                    int receivedAckableMessageCount = receivedAckableMessages.size();
+                    if (receivedAckableMessageCount < 1) {
+                        return Mono.empty();
+                    }
+
+                    // Consume and ack messages.
+                    return pullResponseConsumer
+                            .accept(pullResponse)
+                            .checkpoint("pullResponseConsumer")
+                            .flatMap(ackRequest -> client
+                                    .ack(config.getProjectName(), config.getSubscriptionName(), ackRequest)
+                                    .thenReturn(new PubsubSubscriberConsumption(pullResponse, ackRequest)));
+
+                })
+                .checkpoint("consumeAndAck");
+    }
+
+    private <V> Mono<V> rateLimit(Mono<V> mono) {
+        return mono
+                .flatMap(value -> {
```

Comment:
I would use `.delayUntil` and `return Mono.delay` / `return Mono.empty` in such places, should have less overhead